### DISCUSSION TITLE
Error with wasm32-unknown-unknown only on github action workers

### DISCUSSION THREAD

**Discussion by nfachan** [ipetkov/crane#515] [Q&A] [answered]
Hello.

I've been scratching my head on this for a while, and I'm not sure what to do. We have some scripts meant for use in CI. They invoke `nix develop` and then run various `cargo` commands. These work well for local development, but fail on GitHub. In particular, the `wasm32-unknown-unknown` target doesn't get installed on GitHub workers.

[Here's the flake.nix](https://github.com/maelstrom-software/maelstrom/blob/5ee29932792306aab0586b5b307bb478daf4e6d8/flake.nix).
[Here's a build step where](https://github.com/maelstrom-software/maelstrom/actions/runs/7803979006/job/21284952500) the `wasm32-unknown-unknown` target isn't available.
[Here's a build step where](https://github.com/maelstrom-software/maelstrom/actions/runs/7804086024/job/21285294075) we use vanilla `nix build` and it does work.

[This is what](https://github.com/maelstrom-software/maelstrom/blob/main/scripts/wrapper.sh) our wrapper script looks like.
[This is what](https://github.com/maelstrom-software/maelstrom/blob/main/scripts/lint-inner.sh) an inner script looks like.

**Comment by ipetkov** [comment]
I'm not entirely certain what is happening there. Here's a few points which may be of interest:

> error[E0463]: can't find crate for `core`
> error[E0463]: can't find crate for `compiler_builtins`

If any part of your build is requiring the sources of stdlib crates to be built, you may need to add `extensions = [ "rust-src" ];` to your rustToolchain. Take a look at https://crane.dev/examples/build-std.html for more info

---

I noticed that [this build script](https://github.com/maelstrom-software/maelstrom/blob/5ee29932792306aab0586b5b307bb478daf4e6d8/crates/maelstrom-web/build.rs#L123-L138) is running a cargo build from inside of a cargo build. Not sure if there could be some funny business there (maybe the inner invocation is not inheriting arguments/environment variables properly?).

I would consider debugging that invocation directly in your CI and see if anything interesting pops up...

---

Hope this helps!

**Comment by nfachan** [comment]
Thanks for the thoughtful reply.

Upon more digging, I've found that -- for a reason I've not yet determined -- the PATH passed to the build script has `~/.cargo/bin` prepended to it in CI, and that path has a `cargo` binary in it. So, something interesting did pop up.

Now I have to figure out why cargo is doing that!

**Comment by nfachan** [comment, marked as answer]
I haven't yet gotten to the bottom of why cargo (it seems) is prepending to the PATH. But I was able to solve the issue by removing the .cargo/bin directory in CI.

Thanks for your help!
